### ISSUE TITLE
Paket Support

### ISSUE THREAD

**Issue by slang25** [NuKeeperDotNet/NuKeeper#601]
I would love for NuKeeper to support projects that use Paket as their package manager.

There are 3 files that you will typically find in a solution that uses Paket:
- `paket.dependencies` - This is usually solution level, it has the concept of "groups" and a number of dependencies, dependencies can be things other than nuget packages, but here we'd only pick out nuget deps. In this file version ranges are expressed.
- `paket.references` - This is per project and is how you decide what gets included in a project, version numbers are not defined here, you can only pick from packages that were defined in `paket.dependencies`.
- `paket.lock` - This is a lock file that has the resolved version numbers of the packages in `paket.dependencies` and the resolved transitive dependencies. In some cases it could be that this is the only file that is changed during an update (i.e. this file is invalidated).

I'm proposing that NuKeeper ignores `paket.references`, looks at `paket.dependencies`, and for all groups picks out nuget dependencies and updates them, then finally invalidates `paket.lock` (regardless if there was no change to the deps file) and regenerates it.

I am in the process of building this now, hopefully will have something over the weekend.

**Comment by slang25** [comment]
I don't know why I wrote all that when I could have just included this image which makes it much clearer:
![Paket files](https://fsprojects.github.io/Paket/img/paket-overview.png)

**Comment by slang25** [comment]
How should we be treating lock files with NuKeeper?

Lock files are a thing in Paket, and will be used more in nuget with the arrival of this:
https://github.com/NuGet/Home/wiki/Enable-repeatable-package-restore-using-lock-file

Do we ignore them in the updating process, then regenerate them (regardless of if there were updates), and in the case where the lock file changes create a PR?

**Comment by AnthonySteele** [comment]
I expect that the lock file would be under source control? therefor, regenerating it and committing any changes would be correct?

**Comment by sanderaernouts** [comment]
@AnthonySteele yes the lock file has to be under source control. Normally you would run [paket restore](https://fsprojects.github.io/Paket/paket-restore.html) on your CI server and locally to get the exact versions of the packages that are stored in the `paket.lock` file. So this file is essential to the paket flow.

A developer will run [paket update](https://fsprojects.github.io/Paket/paket-restore.html) or [paket install](https://fsprojects.github.io/Paket/paket-install.html) manually to update the packages and thus update the `packet.lock` file. Those changes will then again be committed to source control.

It is a convention to have paket.exe available in the repo in a `.paket` subfolder. As a fallback the latest paket bootstrapper can be downloaded from their Github repo (this is what we do in our builds now). 

I would suggest using the [paket outdated](https://fsprojects.github.io/Paket/paket-outdated.html) to find packages that can be updated and the `paket update` command to do the actual updating. You can run [paket update](https://fsprojects.github.io/Paket/paket-update.html) for specific packages or use regular expression in combination with the `--filter` argument. This calculates the new dependency tree and store it in the `paket.lock` file which can then again be commited to source control. This way the paket CLI will handle all the updating.

**Comment by MarcBruins** [comment]
Would be best imo to just create a wrapper around the `.paket.exe` and include the reference in the project just like we do with NuGet. 

We can then take the output from the `paket outdated` and pass that through the NuKeeper system which has all the nice options for it(exclude, include, prerelease etc) and use that to update the `paket.lock` file. Then that `paket.lock` file can be commit as the new change just like we do now.

If we can do it this way we can eventuallyuse the same approach for when the not yet release `nuget outdated` functionallity(https://github.com/NuGet/Home/wiki/Show-outdated-packages).

**Comment by sanderaernouts** [comment]
@MarcBruins 
> As just discussed with @sanderaernouts the best option would imo be to just create a wrapper around the .paket.exe and include the reference in the project just like we do with NuGet.

I suggest using [paket.bootstrapper.exe](https://fsprojects.github.io/Paket/bootstrapper.html) and renaming it to `paket.exe`. This will enable allow the bootstrapper to download the right version of Paket either by using [magic mode](https://fsprojects.github.io/Paket/bootstrapper.html#Magic-mode) or by using [required version](https://fsprojects.github.io/Paket/dependencies-file.html#Required-Paket-version) in the `paket.dependencies` file. Which is the suggested way of working with Paket as outlined in this issue: https://github.com/fsprojects/Paket/issues/3430

**Comment by slang25** [comment]
`groupNameFilter` is just the group name to work on (or `None` if operating on all). Paket has the idea of groups, a bit like how npm does (with `devDependencies` vs `dependencies`) where you might have some packages for dev purposes or build scripts, and some for your actual shippable code.

It might be a little easier using the `PublicAPI` type, which is just 1 layer up from the functions you've linked to, they are easier to consume and are intended for this use (I think we might be the first to consume it after searching GitHub).

https://github.com/fsprojects/Paket/blob/2790ed3c57ccbb34c220b21b96b587d2837586e0/src/Paket.Core/PublicAPI.fs#L429
https://github.com/fsprojects/Paket/blob/2790ed3c57ccbb34c220b21b96b587d2837586e0/src/Paket.Core/PublicAPI.fs#L308

**Comment by MarcBruins** [comment]
Let's discuss on Gitter. I can't acces those `PublicApi` methods..

**Comment by MarcBruins** [comment]
@sanderaernouts and made a failed attempt yesterday to add `Paket` to NuKeeper. The reason being is that Nuget is very much woven throughout the `LocalEngine`(we started with the Update command). We've tried to create a IUpdateFinder implemenation for `Paket` but that is already an interface for nuget. For example we have to return a `PackageUpdateSet` which does not map to how `Paket` does updates.

To succesfully implemented `Paket` we have to create a new `PaketEngine` which handles the flow for paket. Which shouldn't be that hard because we use the `Paket.Core` nuget package which provides a lot of functionality.
